### PR TITLE
Use static_cast rather than reinterpret_cast whenever possible

### PR DESCRIPTION
**Pull request by vjpai**
In particular, you should use `static_cast` for going up or down a class hierarchy or for converting between `T*` and `void*` .

Many of the remaining uses of `reinterpret_cast` are for things that are "morally" part of a class hierarchy (e.g., `grpc_endpoint` vs `mock_endpoint`) but currently aren't because they are expressed in C-style. This will allow further reduction of `reinterpret_cast` in the future.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: test/core/util/trickle_endpoint.cc [hunk 1]
```diff
@@ -132,7 +132,7 @@ static int te_get_fd(grpc_endpoint* ep) {
 }
 
 static void te_finish_write(void* arg, grpc_error* error) {
-  trickle_endpoint* te = static_cast<trickle_endpoint*>(arg);
+  trickle_endpoint* te = reinterpret_cast<trickle_endpoint*>(arg);
```
Comment (dgquintas) [inline test/core/util/trickle_endpoint.cc]: If `static_cast` was compiling here, why switch to `reinterpret_cast`?

Comment (vjpai) [reply test/core/util/trickle_endpoint.cc]: Oops, I slipped. Thanks for noticing.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/core/util/trickle_endpoint.cc (hunk 1 above)
Comment (vjpai) [reply test/core/util/trickle_endpoint.cc]: Specifically, my method here was to `s/reinterpret_cast/static_cast/g`, get all the build errors and change those back with an Emacs macro. Looks like I slipped here. I'll check the others.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
Thanks for doing this Vijay! From the errors, it looks like some of the conversions weren't part of a hierarchy or void * to and fro conversions.

LGTM once it builds :)

**Comment by vjpai** [comment]
I wasn't building Cronet until now. Hopefully this goes through now.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]      +8  +0.0%

  [ = ]       0 TOTAL       +8  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]      +8  +0.0%

  [ = ]       0 TOTAL       +8  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by vjpai** [comment]
Rebased, diff unchanged

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]      +8  +0.0%

  [ = ]       0 TOTAL       +8  +0.0%



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**